### DISCUSSION TITLE
Help wanted: VM/Hypervisor hangs on boot when a tap network interface is configured

### DISCUSSION THREAD

**Discussion by nagisa** [cloud-hypervisor/cloud-hypervisor#5733] [General]
**Describe the bug**

When I boot a systemd-based VM with a TAP network interface enabled, I find that somewhere along the boot the VM stops making progress. I see 100% CPU usage on the hypervisor process. The guest linux kernel sometimes prints

```
[   27.438419] rcu: INFO: rcu_preempt detected stalls on CPUs/tasks:
[   27.439045] rcu: All QSes seen, last rcu_preempt kthread activity 21002 (4294694586-4294673584), jiffies_till_next_fqs=3, root ->qsmask 0x0
[   27.440236] rcu: rcu_preempt kthread starved for 21002 jiffies! g20361 f0x2 RCU_GP_WAIT_FQS(5) ->state=0x0 ->cpu=0
[   27.441257] rcu:     Unless rcu_preempt kthread gets sufficient CPU time, OOM is now expected behavior.
[   27.442136] rcu: RCU grace-period kthread stack dump:
[   27.442797] rcu: Stack dump where RCU GP kthread last ran:
```

If I boot this same VM with `qemu` or `crosvm`, using exactly the same guest and equivalent hypervisor settings, the guest boots successfully. This suggests to me that an issue is more likely in `cloud-hypervisor`, but I cannot really tell for sure. Similarly if I just remove the `--net` argument the guest boots with `cloud-hypervisor` just as well.

Unfortunately I don’t really have the knowledge necessary to figure out *how* to investigate this further, and hence the help request.

**To Reproduce**
Steps to reproduce the behaviour: Will try to come up with something if there is no way for me to investigate this by myself.

**Version**

Output of `cloud-hypervisor --version`: 33.0

Did you build from source, if so build command line (e.g. features):

**VM configuration**

What command line did you run (or JSON config data):

```
sudo /nix/store/28wd1f0qndyd9x5p410b96lxjggphjmq-cloud-hypervisor-33.0/bin/cloud-hypervisor --cpus boot=1 --watchdog --console null --serial tty --kernel /nix/store/w3shfnmgpl94jmg9byv7mdsx2gp5kacl-rust-hypervisor-firmware-x86_64-elf-0.4.2/bin/hypervisor-fw --cmdline console=ttyS0 reboot=t panic=-1 init=/nix/store/xa83ssy8qqjs3b1hgn32b7sclpqjc79p-nixos-system-unifi-netserv-23.11.20230817.42c2560/init --seccomp true --memory mergeable=on,shared=on,size=1024M --disk path=/nix/store/0j5162xrmqkjfg6zlw31kcd2wv69j1a6-microvm-bootdisk.img,readonly=on --fs tag=unifi-netserv-var-lib-unifi-data,socket=unifi-netserv-virtiofs-unifi-netserv-var-lib-unifi-data.sock --fs tag=journal,socket=journal.sock --fs tag=ro-store,socket=unifi-netserv-virtiofs-ro-store.sock --api-socket unifi-netserv.sock --net tap=unifi-netserv0,mac=00:02:00:01:01:01
```

On the host side I created the TAP interface as such:

```
sudo ip tuntap add unifi-netserv0 mode tap user root
```

I also tried `--net tap=,mac=,ip=,mask=` as presented in the README as well as `iommu=on/off` but the result was the same in either case.

Guest OS version details: An "empty" NixOS 23.11 system built from sources from 2 weeks ago or so
Host OS version details: NixOS 23.11 system built from sources from 2 weeks ago

In either case linux kernel is 6.1.46.

**Logs**

Output of `cloud-hypervisor -v` from either standard error or via `--log-file`: [log.txt](https://github.com/cloud-hypervisor/cloud-hypervisor/files/12521557/log.txt)

Linux kernel output: no logs from the host kernel as far as I can tell.

**Comment by rbradford** [comment]
In the log output (including the serial) it looks like it boots fine - I see that you're using --console null and --serial tty but specifying `console=ttyS0` on the command line - is it possible your getty is running on the virtio-console which is why it's not appearing for you. Use `--console off` to disable that.

**Comment by nagisa** [comment]
> is it possible your getty is running on the virtio-console which is why it's not appearing for you

The thing is – my getty shows up perfectly fine on the serial interface (and hypervisor’s CPU utilization stays at a reasonable level of 0%) if the `--net` option is removed. Presence or absence of `--net` option shouldn't affect how getty works, should it?

Regardless changing `--console null` to `--console off` did not change anything, unfortunately.

I have also looked at the guest’s journal (by putting it on a shared filesystem), but it does not really have anything suspicious either.

All things considered, *if* my deduction that this is a hang somewhere, is correct, then I wouldn't expect systemd or similar logs to show anything terribly wrong either.

**Comment by Geo-7** [comment]
I have the exact same problem
host OS = Alpine Linux edge kernel 6.6.4-1-lts
cloud-hypervisor v36.0.0
```
cloud-hypervisor \
        --kernel ./hypervisor-fw \
        --disk path=jammy-server-cloudimg-amd64.raw \
        --cpus boot=4 \
        --memory size=2048M \
        --net "tap=" \
        --seccomp false \
        --log-file net.log \
        -v
```
It boots fine without --net

**Comment by thomasbarrett** [comment]
I have also experienced this issue when using a virtio-net devices. The cloud-hypervisor integration tests worked on my machine, but using a virtio-net with a stock ubuntu 22.04 image did not. This indicates that the issue is with triggered by specific guest behavior. Perhaps there is a virtio-net feature flag being negotiated that isn't fully supported and/or tested be cloud-hypervisor.

In my case, the guest kernel logs indicated a hang in the virtio-net client code. I don't use virtio-net (and am not entirely familiar with the spec), so I haven't looked into this too deeply.

**Reply by rbradford** [reply]
Yes, this is a known issue in some kernels used on Ubuntu 22.04 - https://github.com/cloud-hypervisor/cloud-hypervisor/issues/5998

**Reply by thomasbarrett** [reply]
👀 ty!

**Comment by mikroskeem** [comment]
I ran into similar issue on both v52.0 and v51.2, can reproduce with [Ubuntu 24.04.2 LTS cloud image](https://cloud-images.ubuntu.com/releases/noble/release-20260321/ubuntu-24.04-server-cloudimg-arm64.img). This seems to only happen on aarch64 for me though... x86_64 works fine.

Using https://github.com/cloud-hypervisor/edk2/releases/download/ch-1e1b96f126/CLOUDHV_EFI.fd for firmware.

```
Press ESCAPE for boot options ....EFI stub: Booting Linux Kernel...
EFI stub: Loaded initrd from LINUX_EFI_INITRD_MEDIA_GUID device path
EFI stub: Generating empty DTB
EFI stub: Exiting boot services...
[    0.000000] Booting Linux on physical CPU 0x0000000000 [0x610f0000]
[    0.000000] Linux version 6.8.0-106-generic (buildd@bos03-arm64-013) (aarch64-linux-gnu-gcc-13 (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0, GNU ld (GNU Binutils for Ubuntu) 2.42) #106-Ubuntu SMP PREEMPT_DYNAMIC Fri Mar  6 07:44:40 UTC 2026 (Ubuntu 6.8.0-106.106-generic 6.8.12)
[    0.000000] KASLR enabled
[    0.000000] efi: EFI v2.7 by EDK II
[    0.000000] efi: ACPI 2.0=0x5f409018 MEMATTR=0x5ec75018 MOKvar=0x5fa50000 INITRD=0x5ecb2f18 RNG=0x5f409998 MEMRESERVE=0x5ecb2e18
...
[    2.242660] Run /init as init process
Loading, please wait...
Starting systemd-udevd version 255.4-1ubuntu8.12
[    2.318939] virtio_net virtio2 ens3: renamed from eth0
[   12.627797] rcu_tasks_wait_gp: rcu_tasks_trace grace period number 1 (since boot) is 10128 jiffies old.
[   42.875812] rcu_tasks_wait_gp: rcu_tasks_trace grace period number 1 (since boot) is 40376 jiffies old.
Timed out for waiting the udev queue being empty.
[  132.883797] rcu_tasks_wait_gp: rcu_tasks_trace grace period number 1 (since boot) is 130384 jiffies old.
[  246.931874] INFO: task kworker/0:1:10 blocked for more than 122 seconds.
[  246.942024]       Not tainted 6.8.0-106-generic #106-Ubuntu
[  246.950109] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[  246.961986] task:kworker/0:1     state:D stack:0     pid:10    tgid:10    ppid:2      flags:0x00000008
[  246.975502] Workqueue: events do_free_init
[  246.981807] Call trace:
[  246.985420]  __switch_to+0xd8/0x120
[  246.990797]  __schedule+0x2a4/0x788
[  246.996238]  schedule+0x40/0x168
[  247.001290]  schedule_timeout+0x1c8/0x1e0
[  247.007544]  wait_for_completion+0xe8/0x190
[  247.013995]  __wait_rcu_gp+0x1c0/0x228
[  247.019784]  synchronize_rcu+0x178/0x1a0
[  247.025876]  do_free_init+0x44/0xa8
[  247.031627]  process_one_work+0x17c/0x448
[  247.038289]  worker_thread+0x360/0x480
[  247.044539]  kthread+0xf8/0x110
[  247.049296]  ret_from_fork+0x10/0x20
[  247.055031] INFO: task kworker/u4:0:11 blocked for more than 123 seconds.
[  247.065981]       Not tainted 6.8.0-106-generic #106-Ubuntu
[  247.074874] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[  247.087229] task:kworker/u4:0    state:D stack:0     pid:11    tgid:11    ppid:2      flags:0x00000008
[  247.102011] Workqueue: events_unbound fsnotify_mark_destroy_workfn
[  247.111637] Call trace:
[  247.115490]  __switch_to+0xd8/0x120
[  247.121109]  __schedule+0x2a4/0x788
[  247.126631]  schedule+0x40/0x168
[  247.131883]  schedule_timeout+0x1c8/0x1e0
[  247.138153]  wait_for_completion+0xe8/0x190
[  247.144977]  __synchronize_srcu+0x9c/0x100
[  247.151736]  synchronize_srcu+0xc8/0x150
[  247.158094]  fsnotify_mark_destroy_workfn+0x84/0x128
[  247.166000]  process_one_work+0x17c/0x448
[  247.172144]  worker_thread+0x360/0x480
[  247.177920]  kthread+0xf8/0x110
[  247.183063]  ret_from_fork+0x10/0x20
[  247.188748] INFO: task kworker/u4:2:54 blocked for more than 123 seconds.
[  247.199335]       Not tainted 6.8.0-106-generic #106-Ubuntu
[  247.208130] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[  247.220401] task:kworker/u4:2    state:D stack:0     pid:54    tgid:54    ppid:2      flags:0x00000008
[  247.235554] Workqueue: events_unbound fsnotify_connector_destroy_workfn
[  247.246309] Call trace:
[  247.250113]  __switch_to+0xd8/0x120
[  247.255398]  __schedule+0x2a4/0x788
[  247.261089]  schedule+0x40/0x168
[  247.266059]  schedule_timeout+0x1c8/0x1e0
[  247.272335]  wait_for_completion+0xe8/0x190
[  247.278821]  __synchronize_srcu+0x9c/0x100
[  247.285436]  synchronize_srcu+0xc8/0x150
[  247.291938]  fsnotify_connector_destroy_workfn+0x44/0x88
[  247.300225]  process_one_work+0x17c/0x448
[  247.306932]  worker_thread+0x360/0x480
[  247.313091]  kthread+0xf8/0x110
[  247.317895]  ret_from_fork+0x10/0x20
[  247.323262] INFO: task (udev-worker):102 blocked for more than 123 seconds.
[  247.334247]       Not tainted 6.8.0-106-generic #106-Ubuntu
[  247.343168] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
```
